### PR TITLE
Feature/redirect back or error

### PR DESCRIPTION
**Pull request by martenwallewein**

### CODE REVIEW

File: chrome/popup.js [hunk 1]
```diff
@@ -117,11 +142,127 @@ function deleteHosts(hostlist){
...
+          });
+*/
+
+var getRequestsDatabaseAdapter;
+async function loadRequestInfo() {
+  const databaseAdapter = await getRequestsDatabaseAdapter();
+
+
+  const domainList = document.getElementById("domainlist");
+  const checkedDomains = [];
```
Comment (JordiSubira) [inline chrome/popup.js:110]: Some times not all domains are shown in the popus, I wonder if it has to do with how and when this checkedDomains set is (re-)created.

File: chrome/popup.js [hunk 2]
```diff
@@ -1,30 +1,37 @@
...
-        "www.scionlab.chat",
-        "www.scion-pathguess.game"
-      ];
-      saveStorageValue('list', [...hostSet]).then(() => {
-        console.log('Initialized hosts');
-      })
-    }
-    displayHostList(hostSet); 
-  });
+  // getStorageValue('list').then((hostSet) => {
```
Comment (JordiSubira) [inline chrome/popup.js]: Remove?

File: chrome/options.js [hunk 3]
```diff
@@ -60,8 +88,123 @@ async function applyWhitelist(isd, checked){
...
+
+function toggleGlobalStrictMode() {
+    toggleGlobalStrict.checked = !toggleGlobalStrict.checked;
+    if (toggleGlobalStrict.checked) {
+        lineStrictMode.style.backgroundColor = '#48bb78';
+    } else {
+        lineStrictMode.style.backgroundColor = '#cccccc';
+    }
+    saveStorageValue('globalStrictMode', toggleGlobalStrict.checked).then((val) => {
+
```
Comment (JordiSubira) [inline chrome/options.js]: Does this then() clause do anthing useful?

File: chrome/database.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+}
+
+export const getRequestsDatabaseAdapter = () => {
+    return new Promise((resolve) => {
+        resolve(new DatabaseAdapter("requests"));
+    });
+}
+
+// TODO: For later, if we want to persist things, but for now
```
Comment (JordiSubira) [inline chrome/database.js]: Is this a leftover comment? Remove  it if that's the case

File: chrome/database.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+function saveStorageValue(key, value) {
+    return new Promise((resolve, reject) => {
+      chrome.storage.sync.set({[key]: value}, function() {
+        resolve();
+      });
+    });
+  }
+
+  function getStorageValue(key) {
+    return new Promise((resolve, reject) => {
```
Comment (JordiSubira) [inline chrome/database.js]: Remove reject, it isn't used.

File: chrome/database.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+// This is crap. The popup and the browser extension have different local/session storages
+// ...
+
+if(window.database === undefined) {
+    window.database = database;
+}
+
+function saveStorageValue(key, value) {
+    return new Promise((resolve, reject) => {
```
Comment (JordiSubira) [inline chrome/database.js]: Remove reject, it isn't used.

File: chrome/database.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
+
+const database = {
+    requests: [],
+}
+
+// This is crap. The popup and the browser extension have different local/session storages
```
Comment (JordiSubira) [inline chrome/database.js]: Can we briefly explain why do we have this? So that we can change/fix it in the future?

File: chrome/background.js [hunk 8]
```diff
@@ -157,8 +151,177 @@ getStorageValue('extension_running').then(extensionRunning => {
...
+    const url = new URL(details.url);
+    // The actual URL that we need is in ?url=$url
+    const target = url.search.split("=")[1];
+    const targetUrl = new URL(target);
+    knownNonSCION[targetUrl.hostname] = true;
+  }
+}
+
+
+/* Deprecated stuff */
```
Comment (JordiSubira) [inline chrome/background.js]: Remove?

File: chrome/background.js [hunk 9]
```diff
@@ -157,8 +151,177 @@ getStorageValue('extension_running').then(extensionRunning => {
...
+    requestId: requestInfo.requestId,
+    tabId: requestInfo.tabId,
+    domain: url.hostname,
+    mainDomain: requestInfo.initiator ? new URL(requestInfo.initiator).hostname : '',
+  };
+
+
+  // If we don't have any information about scion-enabled or not
+  if (!knownNonSCION[url.hostname] && !knownSCION[url.hostname]) {
+    // TODO: We could do this also in the onErrorOccured/onBeforeRedirect...
```
Comment (JordiSubira) [inline chrome/background.js]: Do we need this if we already have the onErrorOccured/onBeforeRedirect hack?

**Review by JordiSubira — CHANGES_REQUESTED** [review]

File: chrome/popup.js (hunk 1 above)
Comment (martenwallewein) [reply chrome/popup.js:110]: I will check this next. So you can keep the PR open until it's fixed or merge it, what you prefer. But a good indication, thanks!

File: chrome/database.js (hunk 7 above)
Comment (martenwallewein) [reply chrome/database.js]: Done

File: chrome/background.js (hunk 9 above)
Comment (martenwallewein) [reply chrome/background.js]: This was quite interesting. I tried to do it in this hack but it didn't work. I wrote a comment why.

File: chrome/popup.js (hunk 1 above)
Comment (martenwallewein) [reply chrome/popup.js:110]: I think this is solved. We should merge this branch before HotNets if we want to reference the code in the paper

File: chrome/database.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        this.table = table;
+    }
+
+    persist = () => {
+        return debounce(() => {
+            save();
+        })
+    }
+
+    // This part was the reason for missing SCION domain indicators.
```
Comment (JordiSubira) [inline chrome/database.js:62]: I tested it out and works fine for me. IMO it is an acceptable solution for now even the persistance logic seems quite complex for what we are doing so far. I think first we need a clear strategy how to proceed with the extension development.

**Review by JordiSubira — APPROVED** [review]

File: chrome/database.js (hunk 10 above)
Comment (martenwallewein) [reply chrome/database.js:62]: I definitely agree to this!

**Pull request merged**